# Images
## This notebook outlines the techniques to be used in working with images using OpenCV library

## Learning Outcomes
### - Load an image
### - Grab details of a loaded image
### - Display an image
### - Saving an image
### - Drawing Shapes

### Import the necessary libraries

In [22]:
#!pip install opencv-python

In [24]:
from __future__ import print_function
import argparse
import cv2

### Load an image
#### imread( )

In [27]:
image = cv2.imread("image.jpg")

What's this?
- Numpy array

### Let's get into the image details
- Shape

In [31]:
image.shape

(350, 350, 3)

What's this?

- 228
- 350
- 3

### Write a function that takes image as an input and outputs the width, height and number of channels in the image

In [36]:
def get_details(image_file):
    image = cv2.imread(image_file)
    width = image.shape[1]
    height = image.shape[0]
    channels = image.shape[2]
    return width, height, channels

In [38]:
width, height, channels = get_details("image.jpg")
print(f"Width = {width} \nHeight = {height} \nChannels = {channels}")

Width = 350 
Height = 350 
Channels = 3


### Display an image
#### imshow( )

In [41]:
cv2.imshow("Image", image)
cv2.waitKey(0)

-1

### Save an image
#### imwrite( )

In [43]:
cv2.imwrite("newimage.jpg", image)

True

In [45]:
! ls

'ls' is not recognized as an internal or external command,
operable program or batch file.


### Load the newly created image

In [47]:
new_image = cv2.imread("newimage.jpg")
new_image.shape

(350, 350, 3)

In [49]:
cv2.imshow("NewImage", new_image)
cv2.waitKey(0)

-1

### Let's understand the image
- Pixel: atomic unit of an image

In [51]:
(b, g, r) = image[10, 10]
print("Pixel at (10, 10) - Red: {}, Green: {}, Blue: {}".format(r, g, b))

Pixel at (10, 10) - Red: 242, Green: 233, Blue: 228


### Change the value of the pixel

In [54]:
image[10, 10] = (0, 0, 255)
(b, g, r) = image[10, 10]
print("Pixel at (10, 10) - Red: {}, Green: {}, Blue: {}".format(r, g, b))

Pixel at (10, 10) - Red: 255, Green: 0, Blue: 0


In [56]:
cv2.imshow("Changed Image", image)
cv2.waitKey(0)

-1

### Slicing an image
- Pick the top-left corner 100 x 100 pixels

In [58]:
corner = image[0:100, 0:100]
cv2.imshow("Corner", corner)
cv2.waitKey(0)

-1

### Convert them into green color

In [60]:
image[0:100, 0:100] = (0, 255, 0)

In [62]:
cv2.imshow("Updated", image)
cv2.waitKey(0)

-1

## Drawing shapes in image
- Line
- Rectangle
- Circle

### Create a blank canvas (image with 3 channels, 300 x 300 pixels)
- Use Numpy Array

In [66]:
import numpy as np
import cv2
canvas = np.zeros((300, 300, 3), dtype = "uint8")
canvas.shape

(300, 300, 3)

In [68]:
cv2.imshow("Canvas", canvas)
cv2.waitKey(0)

-1

### Let's do a white canvas

In [71]:
white_canvas = cv2.bitwise_not(canvas)
cv2.imshow("White Canvas", white_canvas)
cv2.waitKey(0)

-1

### Draw a line
- Red diagonal line

In [73]:
red = (0, 0, 255)
cv2.line(white_canvas, (0, 0), (300, 300), red)
cv2.imshow("Red Line on White Canvas", white_canvas)
cv2.waitKey(0)

-1

### Draw a red line with fixed width
- Thickness: 5 pixel width

In [75]:
red = (0, 0, 255)
cv2.line(white_canvas, (0, 0), (300, 300), red, 5)
cv2.imshow("Red Thick Line on White Canvas", white_canvas)
cv2.waitKey(0)

-1

### Draw a red 10 x 10 square

In [77]:
white_canvas = cv2.bitwise_not(canvas)
cv2.rectangle(white_canvas, (10, 10), (60, 60), red)
cv2.imshow("Red Square", white_canvas)
cv2.waitKey(0)

-1

### Draw a red 10 x 10 square with 5 pixel width

In [79]:
white_canvas = cv2.bitwise_not(canvas)
cv2.rectangle(white_canvas, (50, 200), (200, 225), red, 5)
cv2.imshow("Red Thick Square Canvas", white_canvas)
cv2.waitKey(0)

-1

### Draw a blue rectangle

In [81]:
white_canvas = cv2.bitwise_not(canvas)
blue = (255, 0, 0)
cv2.rectangle(white_canvas, (200, 50), (225, 125), blue, -1)
cv2.imshow("Blue Rectangle Canvas", white_canvas)
cv2.waitKey(0)

-1

### Draw a white circle at the center of the canvas with increasing radii - from 25 pixels to 150 pixels

In [83]:
canvas = np.zeros((300, 300, 3), dtype = "uint8")
(centerX, centerY) = (canvas.shape[1] // 2, canvas.shape[0] // 2)
white = (255, 255, 255)

for r in range(0, 175, 25):
    cv2.circle(canvas, (centerX, centerY), r, white)

cv2.imshow("Circles Canvas", canvas)
cv2.waitKey(0)

-1

### Draw 25 random circles
- randomly generate a radius size between 5 and 200,
- generate a random color, and then pick a random
- point on our canvas where the circle will be drawn

In [87]:
for i in range(0, 25):
    radius = np.random.randint(5, high = 200)
    color = np.random.randint(0, high = 256, size = (3,)).tolist()
    pt = np.random.randint(0, high = 300, size = (2,))
    cv2.circle(canvas, tuple(pt), radius, color, -1)


cv2.imshow("Random Circles Canvas", canvas)
cv2.waitKey(0)

-1